### ISSUE TITLE
Mechanism in Jetty Client to Jetty Client to know that the underlying TCP Connection is not responding to the messages sent towards the server ?

### ISSUE THREAD

**Issue by sanchit1arora** [eclipse/jetty.project#9070]
**Jetty version**
11.0.11
**Java version**
17.0.1
**Question**
Is there a way for Jetty Client to know that the underlying TCP Connection is not responding to the messages sent towards the server  and terminate that connection and re-establish connection towards the server?

**Comment by sbordet** [comment]
It's not clear what you ask. Make an example.

**Comment by sanchit1arora** [comment]
For Example
 - 
Jetty Client is sending messages to the server using TCP connection. 
If some messages reach the server and we don't seem to receive an ACK for those messages due to network latency/ send queue full at client side, then how does Jetty CLient know if there is a issue with underlying  TCP connection? Hows does it resolve that issue?

**Comment by sbordet** [comment]
The client won't receive a response in time, and it will timeout, either via `Request.timeout()` set by the application, or via idle timeout.

**Comment by owaiscs001** [comment]
Hi @sbordet ,

Today we did some testing on the similar direction where did the following steps

1. Send a HTTP/2 message from Jetty Client to Server
2. TCP KEEP ALIVE is not enabled in either of the side
3. Due to Step 1, A TCP connection got created between Jetty HTTP/2 Client & Server
4. Applied iptables rule to drop all the packets destined for the Jetty HTTP/2 Client TCP port
5. Post that we again sent a HTTP/2 message from Jetty HTTP/2 Client to Server
6. TCPDump shows that packet reached server, but the TCP Ack from the server never reached the Jetty HTTP/2 Client due to iptables rule applied in Step 4 above
7. Post this retransmission attempts were seen from Jetty HTTP/2 Client to Server
8. And any new HTTP/2 request from Jetty HTTP/2 Client never leaves the TCP socket and stays in sendQ only (as confirmed from tcpdump)
9. This behavior continues for all the subsequent HTTP/2 request from Jetty HTTP/2 Client to Server

Wondering , whether there is any way in Jetty to detect this condition and initiate a NEW TCP connection to avoid this Head of the line blocking situation.

For example:
1. Does Jetty allow setting TCP socket options like keep alive etc
2. any other suggestion ?

**Comment by sbordet** [comment]
I don't think you can reliably detect this condition.

The connection will eventually idle timeout, all the requests failed, and the next request after the previous connection has been closed will trigger the opening of a new connection.

Also, in a real world scenario, if the server is so borked, opening a new connection will probably fail (as the server won't send SYN and ACK back), and even if a new connection is opened, chances are that the server won't send back ACKs again, so you will be stuck again.

I don't see how setting TCP's keepalive will help.

**Comment by owaiscs001** [comment]
Hi @sbordet,

By Idle Timeout you mean the HTTP/2 idle timeout right ?
As per our observation, if we keep sending HTTP/2 message, then the idle timeout is not expiring.
To test that, we set the idle timeout as 5 mins and
1. kept sending 1 HTTP/2 packet every one min and the timer never got expired
2. Also, the HTTP/2 Message was never successfully sent, the TCP socket's send-Q kept growing  (1430)
tcp6       0   1430 x.x.x.x:49626    y.y.y.y:8080       ESTABLISHED 403976/java

This is where TCP KeepAlive can help.
If TCP KeeeAlive packets are not acknowledged for the configured no of probes, then it will terminate the connection and that will clean the context.
Hence, for next HTTP/2 request for the same destination, a new TCP Connection will be established.

**Comment by sbordet** [comment]
Ah yes, if you are sending the requests over the network you will not idle timeout.

However, the requests won't be answered, so you can apply a request timeout via:

```java
httpClient.newRequest(...)
  ...
  .timeout(1, SECONDS)
  .send(...)
```

The ability to configure TCP keepalive is available from Java 11, but they are available in the `jdk.net` module, so:

```java
ClientConnector connector = new ClientConnector() {
  protected void configure(SelectableChannel selectable) throws IOException {
    super.configure(selectable);
    if (selectable instanceof NetworkChannel) {
      NetworkChannel channel = (NetworkChannel)selectable;
      channel.setOption(jdk.net.ExtendedSocketOptions.TCP_KEEPINTERVAL, 2); // seconds
    }
  }
};
```

Let us know if this works for you.

**Comment by owaiscs001** [comment]
Hi @sbordet,

Thanks for the above solution.

1. We were able to configure TCP keepalive, But the problem with tcp-retransmissions still persists.
2. Tcp-retransmissions are happening till 15 min and no keepalive segment is observed in the given time window.
3. Then after 15 min the connection is taken down.

Can you provide some direction on how this time window/Tcp-retransmissions count can be controlled by jetty.

**Comment by sbordet** [comment]
TCP keepalives are sent when the connection is idle, and only after `TCP_KEEPIDLE` which defaults to 2 hrs.
And the connection is closed by the TCP keepalive mechanism after `TCP_KEEPCOUNT * TCP_KEEPINTERVAL`.
With the default values, your connection will be closed after `7200 + 9 * 75 = 7875 s` or about 2 h 11 m. And only if it's idle.

In your case, the connection is not idle as you send requests, so I'm not sure TCP keepalives are sent at all.
Do you see them?

If you see them, the 15 m could be due to the `9 * 75` retries which amount to 11 mins and 15 s, close to 15 m.

How did you exactly configure the TCP keepalive feature?

I would also try another solution, which is to do the sanity check of the connection at the application level, rather than at the OS level. In this way you control everything at the application level, rather than relying on obscure OS settings.
You can use Jetty's low-level `HTTP2Client` and send a PING frame to the server every once in a while.
If you don't see a reply, you know that the connectivity with the server is not good, and you can close all connections.

**Comment by owaiscs001** [comment]
Hi @sbordet,

Tcp keepalive changes always apply on new connections when updated.
Existing connection runs on old tcp keepalive configuration.
I have one additional question: Is it possible to configure tcp keepalive on existing connections in jetty.
If yes, can you please provide some direction on that.

Thanks

**Comment by sbordet** [comment]
> I have one additional question: Is it possible to configure tcp keepalive on existing connections in jetty.
> Tcp keepalive changes always apply on new connections when updated.

What use would you have to change the TCP keepalive on existing connections, if they are not applied?

**Comment by owaiscs001** [comment]
We wanted to keep the TCP keepalive configurable (Even though someone will rarely use it). But if it is changed can those changes be applied to existing connections/sockets as well?

**Comment by sbordet** [comment]
> But if it is changed can those changes be applied to existing connections/sockets as well?

Don't know. Can you try and report back if they are?

**Comment by owaiscs001** [comment]
Hi @sbordet 


I tested the following cases:

I kept the TCP keepalive config as "TCP_KEEPIDLE 10, TCP_KEEPINTERVAL 5, TCP_KEEPCOUNT 4" and created a connection, say C1 with some running traffic.
I updated the TCP keepalive config to "TCP_KEEPIDLE 20, TCP_KEEPINTERVAL 5, TCP_KEEPCOUNT 10" and created another connection, say C2.
Now both C1 and C2 have running traffic.
I stopped the traffic on both C1 and C2.
I applied the iptables rule to drop all the packets destined for C1 and C2 in the jetty.

Result:
Connection C1 was closed by following this config "TCP_KEEPIDLE 10, TCP_KEEPINTERVAL 5, TCP_KEEPCOUNT 4"
Connection C2 was closed by following this config "TCP_KEEPIDLE 20, TCP_KEEPINTERVAL 5, TCP_KEEPCOUNT 10"

Question: Is there any way in the jetty to update config "TCP_KEEPIDLE 20, TCP_KEEPINTERVAL 5, TCP_KEEPCOUNT 10" in C1?
